### PR TITLE
[SandboxIR] Add debug checker to compare IR before/after a revert

### PR DESCRIPTION
**Pull request by slackito**
This will help us catch mistakes in change tracking. It's only enabled when EXPENSIVE_CHECKS are enabled.

### CODE REVIEW

File: llvm/lib/SandboxIR/Tracker.cpp [hunk 1]
```diff
@@ -10,12 +10,81 @@
...
+    if (CurrMSIt == Curr.end()) {
+      DifferenceFound = true;
+      dbgs() << "Module " << M->getName() << " not found in current IR.\n";
+      continue;
+    }
+    const ModuleSnapshot &CurrMS = CurrMSIt->second;
+    if (OrigMS.Hash != CurrMS.Hash) {
+      DifferenceFound = true;
+      dbgs() << "Found IR difference in module " << M->getName() << ".\n";
+      dbgs() << "Original:\n" << OrigMS.TextualIR << "\n";
```
Comment (vporpo) [inline llvm/lib/SandboxIR/Tracker.cpp:54]: This is good for now, but my concern is that these module dumps may become really huge when compiling large applications, so getting these textual dumps will become problematic. Could we perhaps do a per-function check, or per-BB check instead?

File: llvm/unittests/SandboxIR/TrackerTest.cpp [hunk 2]
```diff
@@ -1844,3 +1844,41 @@ define void @foo(i32 %arg, float %farg) {
...
+  auto *F = Ctx.createFunction(&LLVMF);
+  auto *BB = &*F->begin();
+  auto It = BB->begin();
+  sandboxir::Instruction *Add0 = &*It++;
+  sandboxir::Instruction *Add1 = &*It++;
+  sandboxir::IRSnapshotChecker Checker(Ctx);
+  Checker.save();
+  Add1->setOperand(1, Add0);
+  EXPECT_DEATH(Checker.expectNoDiff(), "Found IR difference");
+}
```
Comment (vporpo) [inline llvm/unittests/SandboxIR/TrackerTest.cpp:1884]: perhaps we should also exercise running `save()` again (or reset() + save()) after some changes are made?

File: llvm/include/llvm/SandboxIR/Tracker.h [hunk 3]
```diff
@@ -67,6 +67,60 @@ class CmpInst;
...
+  // practice there will always be one module, but sandbox IR save/restore ops
+  // work at the Context level, so we must take the full state into account.
+  using ContextSnapshot = DenseMap<llvm::Module *, ModuleSnapshot>;
+
+  ContextSnapshot OrigContextSnapshot;
+
+  // True if save() was previously called. This helps us distinguish between
+  // "expectNoDiff was called without calling save" and "save was called but
+  // the saved snapshot is empty".
+  bool HasSavedState = false;
```
[the remark this answers is not in the document]
Comment (slackito) [reply llvm/include/llvm/SandboxIR/Tracker.h:98]: The reason, as the comment above says, is that I think the snapshot could be empty. What if we do something like:
```
sandboxir::Context Ctx;
Ctx.save();  // try to save without creating any modules or functions
Ctx.revert();
```
Sure, doing that is not useful, but it doesn't strike me as incorrect usage of the API either.

[the remark this answers is not in the document]
Comment (vporpo) [reply llvm/include/llvm/SandboxIR/Tracker.h:98]: But if we have nothing to check, shouldn't `expectNoDiff()` pass anyway?

If you are trying to check that we ran `save()` before running `revert()` then I think this check shouldn't be the job of the `IRSnapshotChecker`.

File: llvm/unittests/SandboxIR/TrackerTest.cpp (hunk 2 above)
Comment (slackito) [reply llvm/unittests/SandboxIR/TrackerTest.cpp:1884]: Added a test case that runs `save()` twice and checks the second time replaces the first one.

File: llvm/lib/SandboxIR/Tracker.cpp (hunk 1 above)
Comment (slackito) [reply llvm/lib/SandboxIR/Tracker.cpp]: Switched to per-function snapshots. I looked into using BBs instead but StructuralHash only takes Module or Function.

File: llvm/include/llvm/SandboxIR/Tracker.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (slackito) [reply llvm/include/llvm/SandboxIR/Tracker.h]: I was trying to check we ran `save()` before `expectNoDiff` (that is, an invariant at the IRSnapshotChecker level, not at the Context level). But removing it also works, so I just did that.

File: llvm/include/llvm/SandboxIR/Tracker.h [hunk 4]
```diff
@@ -67,6 +67,60 @@ class CmpInst;
...
+  IRSnapshotChecker(Context &Ctx) : Ctx(Ctx) {}
+
+  /// Saves the current state.
+  void save();
+
+  /// Checks current state against saved state, crashes if different.
+  void expectNoDiff();
+
+  /// Empties saved state.
+  void reset();
```
[the remark this answers is not in the document]
Comment (slackito) [reply llvm/include/llvm/SandboxIR/Tracker.h]: No need to call it from `save`, as it will replace the old snapshot with the new one. I've just removed it.

**Review by vporpo — APPROVED** [review]
Looks good, thanks!

**Pull request merged**